### PR TITLE
feat: update e2e tests to accommodate for progressive strategy

### LINKED ISSUE #393: e2e test for Progressive - update functional_test.go to check "promoted" pipeline, monovertex (and isbsvc when ready)
# Summary

We will need to start to test the Progressive logic. 

I'm thinking that we should be able to add "Progressive" as one variety of the current `functional_test.go`. Just like `ppnd`, we'd have some logic in there that's conditionally performed only in the case of `progressive`.

In order to make this work, we need to change how we do things. Instead of looking for a pipeline, monovertex, isbsvc by name, we will look for the one whose upgrade state Label is "promoted" (note the "isbsvc" code isn't in there yet but I am in process with it).

For "nostrategy" and "pause-and-drain", there is only one "promoted" child so it will effectively do the same thing as now except that the test doesn't specify the child resource by name. 

For "progressive", ultimately, the "upgrading" pipeline/monovertex/isbsvc will become the new "promoted" one. 

Please let me know if you see any reason why this won't work.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by dpadhiar**
Fixes #393 

### Modifications

Modifies `functional_test.go` to instead use a helper function `getPipelineName` or `getMonoVertexName` to get the child resource's name instead of outright using the constant we have previously used before. This is due to the Progressive strategy of updating rollouts eventually changing the instance of each child resource ie. `my-pipeline-0` will be `my-pipeline-1` after a successful update. 

We can achieve this currently by listing the resources with the label `numaplane.numaproj.io/upgrade-state=promoted` which there should only be 1 of with the correct label `numaplane.numaproj.io/parent-rollout-name`.

### Verification

Running `make test-e2e` locally and passing multiple times. Double checking with CI now.

### Backward incompatibilities

No compatibility errors as this is for end to end testing.

### CODE REVIEW

File: tests/e2e/monovertex.go [hunk 1]
```diff
@@ -19,6 +20,27 @@ import (
...
+		unstructList, err := dynamicClient.Resource(getGVRForMonoVertex()).Namespace(namespace).List(ctx, metav1.ListOptions{LabelSelector: label})
+		if err != nil {
+			return false
+		}
+		if len(unstructList.Items) == 0 {
+			return false
+		}
+		monoVertexName = unstructList.Items[0].GetName()
+		return true
+	}, 60*time.Second, testPollingInterval).Should(BeTrue())
```
Comment (juliev0) [inline tests/e2e/monovertex.go:38]: Maybe we don't need `Eventually` here anymore? Are all the calls to here within their own Eventually or Consistently block? Are all of them followed by a `Get()`? If so, could this just return the MonoVertex itself, and then the subsequent `Get()` doesn't need to be called?

Comment (juliev0) [reply tests/e2e/monovertex.go:38]: (same comment for Pipeline)

Comment (dpadhiar) [reply tests/e2e/monovertex.go:38]: Within `monovertex.go` we are either passing the name to another function or calling Get. In `functional_test.go`, we do need to get the Name to check if the MonoVertex is deleted or not.

Comment (juliev0) [reply tests/e2e/monovertex.go:38]: How about in the cases in which we are subsequently calling `Get()`, don't call `Get()` anymore and instead call `getMonoVertex()` which does not contain an Eventually call and just returns the `*Unstructured`? (and do let me know if I'm missing something)

Comment (dpadhiar) [reply tests/e2e/monovertex.go]: Got it! Should be addressed in my latest commit now.

File: tests/e2e/functional_test.go [hunk 2]
```diff
@@ -37,11 +37,11 @@ import (
...
 
 const (
-	isbServiceRolloutName            = "test-isbservice-rollout"
-	pipelineRolloutName              = "test-pipeline-rollout"
-	pipelineName                     = "test-pipeline-rollout-0"
-	monoVertexRolloutName            = "test-monovertex-rollout"
-	monoVertexName                   = "test-monovertex-rollout-0"
+	isbServiceRolloutName = "test-isbservice-rollout"
+	pipelineRolloutName   = "test-pipeline-rollout"
+	// pipelineName                     = "test-pipeline-rollout-0"
```
Comment (dpadhiar) [inline tests/e2e/functional_test.go:42]: Need to remove this and monoVertexName.

File: tests/e2e/pipeline.go [hunk 3]
```diff
@@ -166,12 +169,12 @@ func updatePipelineRolloutInK8S(namespace string, name string, f func(apiv1.Pipe
 	Expect(err).ShouldNot(HaveOccurred())
 }
 
-func getPipelineFromK8S(namespace string, pipelineName string) (*unstructured.Unstructured, numaflowv1.PipelineSpec, numaflowv1.PipelineStatus, error) {
+func getPipelineFromK8S(namespace string, pipelineRolloutName string) (*unstructured.Unstructured, numaflowv1.PipelineSpec, numaflowv1.PipelineStatus, error) {
```
Comment (juliev0) [inline tests/e2e/pipeline.go:172]: nit: just thinking about this function name and the `getPipeline()` name - maybe we can call this one `getPipelineSpecAndStatus()` or something like that>

File: tests/e2e/monovertex.go [hunk 4]
```diff
@@ -88,13 +92,14 @@ func verifyMonoVertexReady(namespace, monoVertexName string) {
 
 }
 
-func verifyMonoVertexStatus(namespace, monoVertexName string, f func(numaflowv1.MonoVertexSpec, kubernetes.GenericStatus) bool) {
+func verifyMonoVertexStatus(namespace, monoVertexRolloutName string, f func(numaflowv1.MonoVertexSpec, kubernetes.GenericStatus) bool) string {
```
Comment (juliev0) [inline tests/e2e/monovertex.go:95]: is this name that it's returning being used anywhere? If not, maybe we can take it out?

Comment (juliev0) [reply tests/e2e/monovertex.go:95]: oh, I see you're using it above

File: tests/e2e/common.go [hunk 5]
```diff
@@ -359,3 +363,19 @@ func closeAllFiles() error {
...
+	unstructList, err := dynamicClient.Resource(gvr).Namespace(namespace).List(ctx, metav1.ListOptions{LabelSelector: label})
+	if err != nil {
+		return nil, err
+	}
+	if len(unstructList.Items) == 0 {
+		return nil, fmt.Errorf("list is empty")
+	}
+
+	return &unstructList.Items[0], nil
+
```
Comment (juliev0) [inline tests/e2e/common.go:380]: good idea to add this one

**Review by juliev0 — APPROVED** [review]

**Pull request merged**